File: csharp/binary-search/Iteration1/BinarySearch.cs
using System;

public static class BinarySearch
{
    public static int Find(int[] input, int value)
    {
        var start = 0;
        var end = input.Length - 1;

        while (start <= end)
        {
            var middle = start + (end - start) / 2;
            var target = input[middle];
            
            if (target == value) return middle;

            if (value > target)
            {
                start = middle + 1;
            }
            else
            {
                end = middle - 1;
            }
        }

        return -1;
    }
}

File: zig/binary-search/Iteration1/binary_search.zig

pub fn binarySearch(comptime T: type, target: T, items: []const T) ?usize {
    if (items.len == 0) return null;
    
    var start: usize = 0;
    var end: usize = items.len - 1;

    while (start <= end) {
        const mid = (start + end) / 2;
        const curr = items[mid];

        if (curr == target) return mid;
        if (mid == 0) return null;

        if (target > curr) start = mid + 1;
        if (target < curr) end = mid - 1;
    }

    return null;
}
